### PR TITLE
Document MainDomKeyDiscriminator & FileSystemMainDomLock

### LINKED ISSUE #3894: Document MainDomKeyDiscriminator setting added in v9.4
https://github.com/umbraco/Umbraco-CMS/pull/12037

### PR DESCRIPTION
**Pull request by p-m-j**
Closes #3894

### CODE REVIEW

**Comment by eshanrnh** [comment]
Thanks for the PR, @p-m-j 🙂💪

We will review this soon🙂

File: Reference/V9-Config/GlobalSettings/index.md [hunk 1]
```diff
@@ -33,6 +33,7 @@ To make it easier to see what values are present in the global section, the foll
       "DisableElectionForSingleServer": false,
       "DatabaseFactoryServerVersion": "SqlServer.V2019",
       "MainDomLock": "MainDomSemaphoreLock",
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:35]: Is this correct? Below `IMainDomLock` is mentioned.

File: Reference/V9-Config/GlobalSettings/index.md [hunk 2]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
 
 ### Main dom lock
 
-This is not a setting that comonly needs to be configured.
+Specifies the implementation of IMainDomLock to be used.
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:139]: Suggestion - Add IMainDomLock in `` as `IMainDomLock`

File: Reference/V9-Config/GlobalSettings/index.md [hunk 3]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
 
-This setting specifies what mechanism will be used for the `MainDomLock`, by default a semaphore lock will be used for when running on Windows, and a sql lock will be used when running on Linux.
+IMainDomLock is used to synchronize access to various resources e.g. Lucene indexes.
 
-This setting is only require if you wish to use a sql lock on a windows machine. Semaphore lock is not supported on Linux.
+Available options:
+
++ **MainDomSemaphoreLock** - Windows only, uses a named Semaphore with a maximumCount of 1 to control acquisition of MainDom status.
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:145]: maximum count

File: Reference/V9-Config/GlobalSettings/index.md [hunk 4]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
++ **SqlMainDomLock** - Available cross platform, uses the database to control acquisition of MainDom status.
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:146]: cross-platform

File: Reference/V9-Config/GlobalSettings/index.md [hunk 5]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
+
+*Available  in v9.4+*
++ **FileSystemMainDomLock** - Available cross platform, uses lock files written to LocalTempPath to control acquisition of MainDom status.
+
+The default selection is platform specific.<br />
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:151]: platform-specific

File: Reference/V9-Config/GlobalSettings/index.md [hunk 6]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
+On Windows MainDomSemaphoreLock will be used unless configured otherwise.<br />
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:152]: On Windows,

File: Reference/V9-Config/GlobalSettings/index.md [hunk 7]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
+On other platforms SqlMainDomLock will be used unless configured otherwise.<br />
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:153]: On other platforms,

File: Reference/V9-Config/GlobalSettings/index.md [hunk 8]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
+
+Additionally SqlMainDomLock will be used when configuration specifies MainDomSemaphoreLock on an unsupported platform.
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:155]: Additionally, SqlMainDomLock will be used when the configuration

File: Reference/V9-Config/GlobalSettings/index.md [hunk 9]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
+
+### Main dom key discriminator
+
+*Available in V9.4+*
+
+For advanced use cases e.g. deployment slot swapping on Azure app services.
+
+When using SqlMainDomLock a MainDomKey is used to identify an instance of a running application.
+
+The MainDomKey is by default comprised of the servers machine name & the application id.
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:165]: server’s

File: Reference/V9-Config/GlobalSettings/index.md [hunk 10]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
+This is generally all that is required to control MainDom status as starting a new process for the same application on the same
+server will result in a matching MainDomKey, requiring that an existing instance yields MainDom status to the new process.
+
+Deployment slots for a given Azure app service share the same machine name and without additional configuration will share a
+MainDomKey and therefore compete for MainDom status, this can be undesirable if attempting to deploy to a deployment slot
+followed by a swap with the production slot as once traffic has switched to the new instance the old production instance reboots
+and can re-aquire MainDom status.
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:172]: re-acquire

File: Reference/V9-Config/GlobalSettings/index.md [hunk 11]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
+See [What happens during a swap](https://docs.microsoft.com/en-us/azure/app-service/deploy-staging-slots#what-happens-during-a-swap).
+
+To prevent this from occurring you can specify a MainDomKeyDiscriminator which should be set as slot specific configuration
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:175]: a slot-specific

File: Reference/V9-Config/GlobalSettings/index.md [hunk 12]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
+to prevent the slots competing for MainDom status.
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:176]: slots from competing

File: Reference/V9-Config/GlobalSettings/index.md [hunk 13]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
+
+It's worth noting that during the swap operation there is a period where both instances will share the same
+configuration and at this point the old instance will yield MainDom status to the new instance.
```
Comment (eshanrnh) [inline Reference/V9-Config/GlobalSettings/index.md:179]: at this point,

**Review by eshanrnh — COMMENTED** [review]
Added couple of suggestions to the grammar.

File: Reference/V9-Config/GlobalSettings/index.md (hunk 1 above)
Comment (p-m-j) [reply Reference/V9-Config/GlobalSettings/index.md:35]: Yep, the configuration value `MainDomLock` controls the implementation of `IMainDomLock`

File: Reference/V9-Config/GlobalSettings/index.md (hunk 5 above)
Comment (p-m-j) [inline Reference/V9-Config/GlobalSettings/index.md]:
```suggestion
The default selection is platform-specific.<br />
```

File: Reference/V9-Config/GlobalSettings/index.md [hunk 14]
```diff
@@ -135,11 +136,47 @@ This setting is used to specify which sql server version that the database is ru
...
+IMainDomLock is used to synchronize access to various resources e.g. Lucene indexes.
 
-This setting is only require if you wish to use a sql lock on a windows machine. Semaphore lock is not supported on Linux.
+Available options:
+
++ **MainDomSemaphoreLock** - Windows only, uses a named Semaphore with a `maximumCount` of 1 to control acquisition of MainDom status.
++ **SqlMainDomLock** - Available cross-platform, uses the database to control acquisition of MainDom status.
+
+*Available  in v9.4+*
++ **FileSystemMainDomLock** - Available cross platform, uses lock files written to LocalTempPath to control acquisition of MainDom status.
```
Comment (p-m-j) [inline Reference/V9-Config/GlobalSettings/index.md:149]:
```suggestion
+ **FileSystemMainDomLock** - Available cross-platform, uses lock files written to LocalTempPath to control acquisition of MainDom status.
```

**Comment by eshanrnh** [comment]
Looks good 🙂Merging it in.

**Review by eshanrnh — APPROVED** [review]

**Comment by Jeavon** [comment]
@p-m-j great work! Is it possible to describe when FileSystemMainDomLock should be used instead of SqlMainDomLock?

**Comment by p-m-j** [comment]
@Jeavon yep but I think I would probably need a bit of careful reflection before I attempted it.

The tl;dr version is
+ pretty much always although it's not as well battle tested yet and there may be gotchas that haven't been accounted for
    + ScheduledPublished now requires a distributed write lock to run so that's now safe if there are temporarily two SchedulingPublishers both with MainDom status.
+ More confident when you know that the processes that will attempt to mutate files guarded by MainDom status all share the same lock files.
+ Very confident for single instance deployments especially on *nix as there's no risk of SQL timeouts and no need to wait for a grace period after signalling to other process to yeild i.e. sqlmaindomlock WaitForExistingAsync/AcquireWhenMaxWaitTimeElapsed
